Migrate label-unapproved-picks from mungegithub to prow

nikhita opened the pull request:
This PR migrates the `label-unapproved-picks` munger from mungegithub to prow. Ref: https://github.com/kubernetes/test-infra/issues/3331.

The original munger did not contain any tests. This PR also adds tests for it.

This PR also makes the plugin configurable:

- The branch name against which PRs should be treated as cherrypicks can now be specified as a regexp.
- The comment that the plugin comments with while adding the `do-not-merge/cherry-pick-not-approved` label can be specified.

This PR also updates the label name from `cherrypick-unapproved` to `cherry-pick-unapproved`.

The content below this line was originally what the PR was based on. This PR contains the following change as well i.e. it updates the comment body to list the patch release managers. 

---

Ref: https://github.com/kubernetes/community/issues/994#issuecomment-370144301. Note: this does not assign the patch release manager, just suggests them so that it's not too noisy for the release manager. Sometimes it's also useful to assign the release manager only after the cherrypick PR has received lgtm + approval.

Part of https://github.com/kubernetes/test-infra/issues/1795

If a cherrypick PR does not have the `cherrypick-approved` label, the comment (by the bot)  just lists that this label does not exist on the PR and adds a `do-not-merge/cherry-pick-not-approved` label. It does not tell us what needs to be done for cherrypick approval. See https://github.com/kubernetes/kubernetes/pull/67616#issuecomment-414436914 for an example.

This PR updates the comment to point to the list of patch release managers so the author of the PR can know whom to assign to get the cherrypick approved.

/sig release
/sig contributor-experience

nikhita commented:
/cc @cblecker @guineveresaenger 
contribex :bat: signal

/cc @tpepper @spiffxp

spiffxp commented:
I get the utility, but I also really don't want any more mungegithub maintenance.  Can we instead port this to a prow plugin?

spiffxp commented:
overhauling cherrypicking is a much larger task, not trying to blow your scope (and I want to discuss this with patch managers)

I think I'm just asking to port this munger

nikhita commented:
@spiffxp :+1:

I'll take a dab at porting it to prow tomorrow (it's 1am here :upside_down_face:)

spiffxp commented:
Really appreciate it.  If it turns out to be too much of a pain, and this is really necessary to move us forward, we can look at merging.  But I think this is the (or very nearly the) last piece of mungegithub to be ported over.  And the bonfire will be glorious.

nikhita commented on mungegithub/mungers/cherrypick-label-unapproved.go:
```diff
@@ -1,117 +0,0 @@
-	obj.WriteComment(labelUnapprovedBody)
-}
-
-func (LabelUnapprovedPicks) isStaleIssueComment(obj *github.MungeObject, comment *githubapi.IssueComment) bool {
```
I'm not sure what this does? I didn't add anything related to this to the plugin for now.

stevekuznetsov replied:
This was the old comment cleanup mechanism

nikhita commented on prow/plugins/cherrypick/cherrypick-unapproved.go:
```diff
@@ -0,0 +1,123 @@
+// Package cherrypick adds the `do-not-merge/cherry-pick-not-approved`
+// label to PRs against a release branch which do not have the
+// `cherrypick-approved` label.
+package cherrypick
```
Should the name be `cherrypick-unapproved` or something like that?

spiffxp replied:
We an [external `cherrypicker` plugin](https://github.com/kubernetes/test-infra/tree/fa801299f10285d4ef491e84bb5f8ce916cf0977/prow/external-plugins/cherrypicker) that implements a `/cherrypick` command so yeah, this package/plugin should be named something else to avoid confusion.

Not sure what off the top of my head though.  Will review and come back

nikhita commented on prow/plugins/cherrypick/cherrypick-unapproved.go:
```diff
@@ -0,0 +1,123 @@
+
+func handlePR(gc githubClient, log *logrus.Entry, pr *github.PullRequestEvent) error {
+	// Only consider the events that indicate opening of the PR
+	if pr.Action != github.PullRequestActionOpened && pr.Action != github.PullRequestActionReopened {
```
PTAL if this right.

stevekuznetsov commented on prow/plugins/cherrypick/cherrypick-unapproved.go:
```diff
@@ -0,0 +1,123 @@
+	)
+
+	// if it is not against a release branch, don't do anything
+	if !strings.HasPrefix(branch, "release-") {
```
Can we please expose this through plugin config?

stevekuznetsov commented on prow/plugins/cherrypick/cherrypick-unapproved.go:
```diff
@@ -0,0 +1,123 @@
+
+	labels, err := gc.GetIssueLabels(org, repo, pr.Number)
+	if err != nil {
+		log.WithError(err).Errorf("failed to list labels on  %s/%s#%d", org, repo, pr.Number)
```
I think we should exit early in this case?

stevekuznetsov commented on prow/plugins/cherrypick/cherrypick-unapproved.go:
```diff
@@ -0,0 +1,123 @@
+
+	// only add the label and comment if none of the approved and unapproved labels are present
+	if err := gc.AddLabel(org, repo, pr.Number, cpUnapprovedLabel); err != nil {
+		log.WithError(err).Errorf("Github failed to add the following label on %s/%s#%d: %s", org, repo, pr.Number, cpUnapprovedLabel)
```
I think we get the repo, pull etc information for free in the logger context so your log messages can elide that information

nikhita commented:
I came across https://github.com/kubernetes/test-infra/issues/8832. From the issue:

> We have a cherrypick-candidate and cherrypick-approved labels and then there is a cherry-pick-not-approved (i.e. do-not-merge/cherry-pick-not-approved). Notice cherrypick vs cherry-pick. Some inconsistency here and it also makes it hard to search issues/labels related to cherry-pick in a single search.

> But I agree we should make them consistent in whatever solution we use to replace the existing cherrypick/cherry-pick process.

I would incline towards `cherry-pick` because the git command is `git cherry-pick`. And there is no mention of a "cherrypick" here: https://git-scm.com/docs/git-cherry-pick.

nikhita commented:
cc @spzala

stevekuznetsov commented on prow/plugins/cherrypickunapproved/cherrypick-unapproved.go:
```diff
@@ -0,0 +1,157 @@
+	labelUnapprovedFormat     = "This PR is not for the master branch but does not have the `%s` label. Adding the `%s` label.\n"
+	assignPatchReleaseManager = `
+	To approve the cherrypick, please assign the patch release manager for the ` + "`%s`" + ` branch by writing ` + "`/assign @username`" + ` in a comment when ready.
+The list of patch release managers for each release can be found [here](https://git.k8s.io/sig-release/release-managers.md).`
```
Oops, missed this -- should be configurable as well for other deployments. Can be done in a follow-up (or by me).

nikhita replied:
> Oops, missed this -- should be configurable as well for other deployments. Can be done in a follow-up (or by me).

Updated in this PR. :)

stevekuznetsov commented on prow/plugins/cherrypickunapproved/cherrypick-unapproved.go:
```diff
@@ -0,0 +1,157 @@
+	// Only the 'Config' and Description' fields are necessary because this
+	// plugin does not react to any commands.
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: "Label `do-not-merge/cherry-pick-not-approved` to PRs against a release branch which do not have the `cherrypick-approved` label",
```
Maybe

```go
 Description: "Label PRs against a release branch which do not have the `cherrypick-approved` label with the `do-not-merge/cherry-pick-not-approved` label.", 
```

?

stevekuznetsov commented on prow/plugins/cherrypickunapproved/cherrypick-unapproved.go:
```diff
@@ -0,0 +1,157 @@
+		Description: "Label `do-not-merge/cherry-pick-not-approved` to PRs against a release branch which do not have the `cherrypick-approved` label",
+		Config: map[string]string{
+			"": fmt.Sprintf(
+				"The cherrypick-unapproved plugin treats PRs against branch names satisfying the regular expression `%s` as cherrypick PRs.",
```
`s/cherrypick/cherry-pick/`

stevekuznetsov commented on prow/plugins/cherrypickunapproved/cherrypick-unapproved.go:
```diff
@@ -0,0 +1,157 @@
+func handlePR(gc githubClient, log *logrus.Entry, pr *github.PullRequestEvent, cp commentPruner, branchRe *regexp.Regexp) error {
+	// Only consider the events that indicate opening of the PR and
+	// when the cherrypick-approved label is added or removed
+	cpApprovedLabelUpdated := (pr.Action == github.PullRequestActionLabeled || pr.Action == github.PullRequestActionUnlabeled) && pr.Label.Name == cpApprovedLabel
```
What if someone manually removes the `do-not-merge/cherry-pick-not-approved` label without adding the approved label?

stevekuznetsov commented on prow/plugins/plugins.go:
```diff
@@ -520,6 +531,9 @@ func (c *Configuration) setDefaults() {
 	if c.Owners.LabelsBlackList == nil {
 		c.Owners.LabelsBlackList = []string{"approved", "lgtm"}
 	}
+	if c.CherryPickUnapproved.BranchRegexp == "" {
```
Can we add a test for this?

nikhita replied:
Done.

nikhita commented:
@stevekuznetsov Addressed your comments. Ptal.

Also, updated the label from `cherrypick-approved` to `cherry-pick-approved`.

However, I did not update the `cherrypick-candidate` label because that's related to the [clearPickAfterMerge munger](https://github.com/kubernetes/test-infra/blob/4876b6c19f7f64ee762e9d036f8ced2dcaf3a724/mungegithub/mungers/cherrypick-clear-after-merge.go#L39) and would involve updates in mungegithub. This munger is not actively deployed too.

cjwagner commented on label_sync/labels.yaml:
```diff
@@ -15,13 +15,13 @@ default:
-      description: Indicates a cherrypick PR into a release branch has been approved by the release branch manager. # Consumed by the kubernetes/kubernetes cherrypick-queue.
-      name: cherrypick-approved
+      description: Indicates a cherry-pick PR into a release branch has been approved by the release branch manager. # Consumed by the kubernetes/kubernetes cherry-pick-queue.
+      name: cherry-pick-approved
```
Please use the `previously` field so that label_sync migrates the old label to the new label.

nikhita replied:
Done.

cjwagner commented on prow/plugins/cherrypickunapproved/cherrypick-unapproved.go:
```diff
@@ -0,0 +1,151 @@
+		if err != nil {
+			return fmt.Errorf("error getting bot name: %v", err)
+		}
+		cp.PruneComments(shouldPrune(log, botName, body))
```
The BotName check is unnecessary, `PruneComments` only considers bot comments already.

cjwagner commented on prow/plugins/plugins.go:
```diff
@@ -459,6 +460,18 @@ type Welcome struct {
+	BranchRe     *regexp.Regexp `json:"-"`
+	// Comment is the comment added by the plugin while adding the
+	// `do-not-merge/cherry-pick-not-approved` label.
+	Comment string `json:"comment,omitempty"`
```
This isn't actually the comment that is added. It is a format string for the comment. I think we should use a literal value here instead. The parameters to the format string are not exposed to the user so it isn't clear how to construct the comment format string. We also don't validate that the provided string is valid.

cjwagner replied:
Yes, having the user provide a format string that must use parameters that aren't exposed to the user isn't good UX. They would need to look at the code to see how the format string is actually used.
We either need to have the user provide a literal string (not a format string) or we need to clearly specify what the format string parameters are and in what order they must appear so that the user can construct their own format string without looking at the code.

nikhita replied:
Ah of course! Fixed to use a literal value. :+1:

cjwagner commented on prow/plugins/cherrypickunapproved/cherrypick-unapproved.go:
```diff
@@ -0,0 +1,141 @@
+}
+
+// shouldPrune finds comments left by this plugin.
+func shouldPrune(log *logrus.Entry, commentBody string) func(github.IssueComment) bool {
```
`log` is not used.
You could just get rid of this entire function now and pass this value to PruneComments instead:
```go
func(comment github.IssueComment) bool {
	return strings.Contains(comment.Body, comment)
}
```

nikhita replied:
Done.

cjwagner commented on prow/plugins/cherrypickunapproved/cherrypick-unapproved.go:
```diff
@@ -0,0 +1,141 @@
+	labels, err := gc.GetIssueLabels(org, repo, pr.Number)
+	if err != nil {
+		log.WithError(err).Errorf("failed to list labels")
+		return err
```
The returned error will be logged so the log message here is redundant.

nikhita replied:
Done.

cjwagner commented:
/hold cancel

nikhita commented:
Cherry-picks are not getting merged right now because the cherry-pick labels were updated from `cherrypick-approved` to `cherry-pick-approved`. The label changes were made across the repos immediately (https://github.com/kubernetes/test-infra/pull/9128#issuecomment-415911123).

However, prow was not bumped. The current prow code looks for a `cherrypick-approved` label and if it doesn't find it, it applies the `do-not-merge/cherry-pick-not-approved` label. Once prow is bumped, the code will look for the `cherry-pick-approved` label and remove the `do-not-merge` label.

Also, even after prow is bumped, all such PRs (https://github.com/kubernetes/kubernetes/labels/cherry-pick-approved) will need a trigger in terms of changing the cherrypick labels or closing/reopening the PR. I'll take care of closing+reopening all the affected ones. :+1:

Can some googler bump prow to get this fixed?

stevekuznetsov commented:
Firstly, thanks @nikhita for the migration!! The work is very appreciated. For the label migration, we can use the `label_sync` tool to migrate one label to another. Hopefully the [doc](https://github.com/kubernetes/test-infra/blob/master/label_sync/README.md) is sufficient to explain how the config should be changed, but otherwise @fejta is a good contact.

stevekuznetsov commented:
(But of course a bump to Prow is also necessary, contact the on-call for that.)

nikhita commented:
> For the label migration, we can use the label_sync tool to migrate one label to another. 

@stevekuznetsov The change to `labels.yaml` was done in this PR and the labels have already been migrated. :) Ref: https://github.com/kubernetes/test-infra/pull/9128#issuecomment-415911123.

> (But of course a bump to Prow is also necessary, contact the on-call for that.)

Have reached out on Slack about this. :+1:
